### ISSUE TITLE
syntax/parse should provide an option to automatically record disappeared uses

### ISSUE THREAD

**Issue by lexi-lambda** [racket/racket#1913]
With the new `~undo` pattern, it’s now possible to create a variant of `~literal` that automatically records disappeared uses with no false positives:

```racket
#lang racket

(require (for-meta 2 racket/base
                     syntax/parse/experimental/template)
         (for-syntax racket/syntax
                     syntax/parse/define)
         syntax/parse/define)

(begin-for-syntax
  (define-syntax ~literal/record
    (pattern-expander
     (syntax-parser
       [(_ lit:id {~optional {~seq #:phase phase:expr}})
        (template {~and {~literal lit {?? {?@ #:phase phase}}}
                        lit-id
                        {~do (define lit-stx (if (syntax-transforming?)
                                                 (syntax-local-introduce #'lit-id)
                                                 #'lit-id))
                             (current-recorded-disappeared-uses
                              (cons lit-stx (current-recorded-disappeared-uses)))}
                        {~undo (current-recorded-disappeared-uses
                                (remove lit-stx (current-recorded-disappeared-uses) eq?))}})]))))

(define-syntax (my-cond stx)
  (with-disappeared-uses
    (syntax-parse stx
      [(_ [test:expr body:expr ...+] ...
          [{~literal/record else} else-body:expr ...+])
       #'(cond
           [test body ...] ...
           [else else-body ...])])))
```

I think this is exciting. Really, though, I want this behavior all the time, and I don’t want to have to manually annotate all my literals with `~literal/record` any more than I want to annotate all my literals with `~literal`.

I believe @rmculpepper at one point said that tracking disappeared uses was out of scope for `syntax/parse`, but I have to ask: why? If absolutely necessary, this could be an opt-in feature by providing a new parse option, `#:track-disappeared-uses`, but I think it makes sense to do this tracking automatically if the result of a `syntax-parse` form is a syntax object. This would make literal tracking *far* more reliable, since it would mostly happen automatically. Currently, missing disappeared use properties seems to be one of the most common mistakes macro authors make due to the way they currently need to be handled manually.

**Comment by rmculpepper** [comment]
Yes, I plan to add this soon. I think what I meant before was that `syntax-parse` can't apply the `'disappeared-uses` syntax-property to the result, because the result might not be syntax. But I agree that tracking the literals matched is a good idea. (But it won't use `current-recorded-disappeared-uses`.)

**Comment by lexi-lambda** [comment]
Can’t `syntax-parse` check if the result is syntax and apply the property if it is? I guess that would mean the final body expression would not be called in tail position, but I’m not sure if that’s a problem or not (and there could always be a `#:no-attach-disappeared-uses` option or something like that to turn it off). It just seems like the right default for the 80% case.

**Comment by AlexKnauth** [comment]
Is the new unwindable state meant to help this?

**Comment by lexi-lambda** [comment]
Closed by 774b02a0b827996506fb6b9396adf5226e68337c.
